File: 1.6 APIs/src/03-retry.ts
import { client } from "./client.js";

const RETRYABLE_STATUS_CODES = new Set([429, 500, 503, 408]);

async function callWithRetry(
    prompt: string,
    maxRetries: number = 3,
): Promise<string> {
    let lastError: any = new Error("Unknown error");

    for (let attempt = 0; attempt < maxRetries; attempt++) {
        try {
            const response = await client.chat.completions.create({
                model: "llama-3.1-8b-instant",
                messages: [{ role: "user", content: prompt }],
                max_tokens: 512,
            });

            return response.choices[0].message.content ?? "";
        } catch (error: any) {
            lastError = error;

            const status = error?.status;

            // Don't retry non-retryable errors
            if (status && !RETRYABLE_STATUS_CODES.has(status)) {
                throw error;
            }


            // On rate limit, respect the Retry-After header if present
            const retryAfter = error?.headers?.["retry-after"]; // API TOLD US TO WAIT, LET'S OBEY
            let delay: number;

            if (retryAfter) {
                delay = parseInt(retryAfter, 10) * 1000; // Convert seconds to ms
            } else {
                // Exponential backoff + jitter
                const base = 1000;                           // 1 second base
                const maxDelay = 30_000;                     // 30 seconds ceiling
                const exponential = base * Math.pow(2, attempt);
                const jitter = Math.random() * 1000;         // 0–1000ms random
                delay = Math.min(exponential + jitter, maxDelay);
            }

            console.log(`Attempt ${attempt + 1} failed (${status}). Retrying in ${Math.round(delay / 1000)}s...`);

            await new Promise(resolve => setTimeout(resolve, delay));

        }
    }

    throw lastError;
}


// Usage
try {
    const result = await callWithRetry("Explain RAG in one sentence.");
    console.log(result);
} catch (err: any) {
    console.error("Call failed:", err);

    // If the thrown value is a plain object (some SDKs throw structured objects),
    // print its enumerable and non-enumerable properties for debugging.
    try {
        const allProps = Object.getOwnPropertyNames(err).reduce((acc: any, k) => {
            acc[k] = (err as any)[k];
            return acc;
        }, {});
        console.error("Error details:", JSON.stringify(allProps, null, 2));
    } catch (e) {
        // ignore
    }

    process.exit(1);
}

File: 1.6 APIs/go/03_retry.go
// 03_retry.go — production calls fail transiently: rate limits (429), overloaded
// servers (500/503), timeouts (408). The fix is to retry, but only on
// retryable statuses, with exponential backoff + jitter, and to obey the
// Retry-After header when the API tells us exactly how long to wait.
package main

import (
	"context"
	"errors"
	"fmt"
	"math"
	"math/rand"
	"strconv"
	"time"
)

var retryableStatus = map[int]bool{429: true, 500: true, 503: true, 408: true}

func callWithRetry(ctx context.Context, c *Client, prompt string, maxRetries int) (string, error) {
	var lastErr error

	for attempt := 0; attempt < maxRetries; attempt++ {
		resp, err := c.Chat(ctx, ChatRequest{
			Model:     "llama-3.1-8b-instant",
			Messages:  []Message{{Role: "user", Content: prompt}},
			MaxTokens: 512,
		})
		if err == nil {
			return resp.Choices[0].Message.Content, nil
		}
		lastErr = err

		// Only retry known-transient API errors. Anything else (bad request,
		// auth failure, network error) is fatal — fail fast.
		var apiErr *APIError
		if !errors.As(err, &apiErr) || !retryableStatus[apiErr.StatusCode] {
			return "", err
		}

		delay := backoff(attempt, apiErr.RetryAfter)
		fmt.Printf("Attempt %d failed (%d). Retrying in %.0fs...\n",
			attempt+1, apiErr.StatusCode, delay.Seconds())

		// Sleep, but wake early if the caller cancels the context.
		select {
		case <-time.After(delay):
		case <-ctx.Done():
			return "", ctx.Err()
		}
	}

	return "", fmt.Errorf("exhausted %d retries: %w", maxRetries, lastErr)
}

// backoff decides how long to wait before the next attempt.
func backoff(attempt int, retryAfter string) time.Duration {
	// The API told us exactly how long to wait — obey it.
	if retryAfter != "" {
		if secs, err := strconv.Atoi(retryAfter); err == nil {
			return time.Duration(secs) * time.Second
		}
	}

	// Otherwise: exponential backoff with jitter, capped at 30s.
	const base = time.Second
	const maxDelay = 30 * time.Second
	exponential := time.Duration(math.Pow(2, float64(attempt))) * base
	jitter := time.Duration(rand.Intn(1000)) * time.Millisecond
	if d := exponential + jitter; d < maxDelay {
		return d
	}
	return maxDelay
}

func retryDemo(ctx context.Context, c *Client) error {
	result, err := callWithRetry(ctx, c, "Explain RAG in one sentence.", 3)
	if err != nil {
		return err
	}
	fmt.Println(result)
	return nil
}
